src/enzyme_ad/jax/Passes/AffineToStableHLORaising.cpp

```diff
@@ -505,6 +513,46 @@ expandAffineExpr(OpBuilder &builder, Location loc, AffineExpr expr,
   llvm_unreachable("unreachable");
 }
 
+/// scope is an operation _in_ the scope we are interested in
+bool isSafeToSpeculativelyExecuteAtScope(Operation *scope, Operation *op) {
+  if (mlir::isPure(op))
+    return true;
+
+  MemRefType ty = nullptr;
+  if (auto read = dyn_cast<affine::AffineReadOpInterface>(op))
+    ty = read.getMemRefType();
+  if (!ty)
+    return false;
+
+  IslAnalysis ia;
+
+  isl_set *array = ia.getMemrefShape(ty);
+  llvm::errs() << "ARRAY\n";
+  isl_set_dump(array);
+
+  auto accessMap = ia.getAccessMap(op);
+  if (!accessMap)
+    return false;
+
+  llvm::errs() << "ACCESSMAP\n";
+  isl_map_dump(accessMap);
```

Comment:
@ivanradanov if there is an inner affine.for would this still work or potential crash?

we definitely want to exclude any if conditions, but we do need to ensure we have all the index vars

Reply:
Currently the use of this function is limited to only this case:

```
`scope` -> affine.if {
             ...
`op` ->      affine.load
```
so no indexes will be gone. I guess I will add a dominance check to make sure we don't misuse it in the future